### PR TITLE
chore: Move AlertmanagerEndpoints validation to pkg/prometheus/server

### PR DESCRIPTION
**Pull request by mviswanathsai**
## Description

This PR moves the logic for storing AM endpoints to the store and the subsequent validation to pkg/prometheus/server as per comments on #6467.

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ x] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification

Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note

```

### CODE REVIEW

**Comment by mviswanathsai** [comment]
I don't understand why the generate and format action is failing.

**Comment by yp969803** [comment]
By looking at the error, the error is 
```
lookup proxy.golang.org: no such host
```  The CI pipeline not able to resolve the DNS,

**Comment by mviswanathsai** [comment]
Right, I don't know if that has anything to do with the changes though.

**Comment by mviswanathsai** [comment]
@simonpasquier could you look into this?

**Comment by slashpai** [comment]
Looks to be sporadic issue, I have re-run the failed job lets see

File: pkg/prometheus/server/operator.go [hunk 1]
```diff
@@ -1221,3 +1222,48 @@ func makeSelectorLabels(name string) map[string]string {
 		"prometheus":                    name,
 	}
 }
+
+func ValidateAlertmanagerEndpoints(am monitoringv1.AlertmanagerEndpoints) error {
```
Comment (slashpai) [inline pkg/prometheus/server/operator.go:1226]: nit: wondering whether we need to expose the method outside the package?

File: pkg/prometheus/server/operator.go [hunk 2]
```diff
@@ -1221,3 +1222,48 @@ func makeSelectorLabels(name string) map[string]string {
...
+	}
+
+	if len(nonNilFields) > 1 {
+		return fmt.Errorf("%s can't be set at the same time, at most one of them must be defined", strings.Join(nonNilFields, " and "))
+	}
+
+	return nil
+}
+
+func AddAlertmanagerEndpointsToStore(ctx context.Context, store *assets.Store, namespace string, ams []monitoringv1.AlertmanagerEndpoints) error {
```
Comment (slashpai) [inline pkg/prometheus/server/operator.go:1252]: same here

File: pkg/prometheus/server/operator.go (hunk 1 above)
Comment (slashpai) [reply pkg/prometheus/server/operator.go:1226]: @simonpasquier wdyt?

File: pkg/prometheus/server/operator.go [hunk 3]
```diff
@@ -1221,3 +1230,48 @@ func makeSelectorLabels(name string) map[string]string {
...
+	}
+
+	return nil
+}
+
+func addAlertmanagerEndpointsToStore(ctx context.Context, store *assets.StoreBuilder, namespace string, ams []monitoringv1.AlertmanagerEndpoints) error {
+	for i, am := range ams {
+		if err := validateAlertmanagerEndpoints(am); err != nil {
+			return fmt.Errorf("alertmanager %d: %w", i, err)
+		}
```
Comment (ArthurSens) [inline pkg/prometheus/server/operator.go:1264]: In `pkg/prometheus/server/operator.go` we already call `validateAlertmanagerEndpoints` right before calling `addAlertmanagerEndpointsToStore`. Do we need to call it again here?

**Review by ArthurSens — COMMENTED** [review]
I believe we still have merge conflicts to solve here 😬

File: pkg/prometheus/server/operator.go (hunk 3 above)
Comment (mviswanathsai) [reply pkg/prometheus/server/operator.go:1264]: I must've messed up the rebase, let me clear it up.

File: pkg/prometheus/server/operator.go [hunk 4]
```diff
@@ -1125,12 +1126,15 @@ func (c *Operator) createOrUpdateConfigurationSecret(ctx context.Context, p *mon
 	}
 
 	if p.Spec.Alerting != nil {
-		for i, am := range p.Spec.Alerting.Alertmanagers {
-			if err := prompkg.ValidateAlertmanagerEndpoints(am, p); err != nil {
+		ams := p.Spec.Alerting.Alertmanagers
+
+		for i, am := range ams {
```
Comment (mviswanathsai) [inline pkg/prometheus/server/operator.go:1131]: Can't recall why I brought this loop (calling `validateAlertmanagerEndpoints()`) outside of `addAlertmanagerEndpointsToStore()`, should I just move it back?

File: pkg/prometheus/server/operator.go [hunk 5]
```diff
@@ -1226,3 +1230,45 @@ func makeSelectorLabels(name string) map[string]string {
...
+		if reflect.ValueOf(v).IsNil() {
+			continue
+		}
+		nonNilFields = append(nonNilFields, fmt.Sprintf("%q", k))
+	}
+
+	if len(nonNilFields) > 1 {
+		return fmt.Errorf("%s can't be set at the same time, at most one of them must be defined", strings.Join(nonNilFields, " and "))
+	}
+
```
Comment (mviswanathsai) [inline pkg/prometheus/server/operator.go:1256]: I am unable to call `validateRelabelConfigs()` inside the relocated `validateAlertmanagerEndpoints()`, what should we do? now that we have both `alertRelabelConfigs` and `relabelConfigs`?

**Review by simonpasquier — APPROVED** [review]

**Pull request merged**